arangod/RocksDBEngine/RocksDBHotBackup.cpp

```diff
@@ -0,0 +1,770 @@
+
+} // RocksDBHotBackup::rebuildPath
+
+
+//
+// @brief Remove file or dir currently occupying "path"
+//
+bool RocksDBHotBackup::clearPath(const std::string & path) {
+  bool retFlag = {true};
+
+  if (basics::FileUtils::exists(path)) {
+    if (basics::FileUtils::isDirectory(path)) {
+      TRI_RemoveDirectory(path.c_str());
+    } else {
+      basics::FileUtils::remove(path);
+    } // else
+
+    // test if still there and error out?
+    if (basics::FileUtils::exists(path)) {
+      retFlag = false;
+      LOG_TOPIC(ERR, arangodb::Logger::ENGINES)
+        << "RocksDBHotBackup::clearPath:  unable to remove previous " << path;
+    } // if
+  } // if
+
+  return retFlag;
+
+} // RocksDBHotBackup::clearPath
+
+
+//
+// @brief Common routine for retrieving parameter from request body.
+//        Assumes caller has maintain state of _body and _valid
+//
+void RocksDBHotBackup::getParamValue(const char * key, std::string & value, bool required) {
+  VPackSlice tempSlice;
+
+  try {
+    if (_body.isObject() && _body.hasKey(key)) {
+      tempSlice = _body.get(key);
+      value = tempSlice.copyString();
+    } else if (required) {
+      if (_valid) {
+        _result.add(VPackValue(VPackValueType::Object));
+        _valid = false;
+      }
+      _result.add(key, VPackValue("parameter required"));
+    } // else if
+  } catch(VPackException const &vexcept) {
+    if (_valid) {
+      _result.add(VPackValue(VPackValueType::Object));
+      _valid = false;
+    }
+    _result.add(key, VPackValue(vexcept.what()));
+  };
+
+} // RocksDBHotBackup::getParamValue (std::string)
+
+
+void RocksDBHotBackup::getParamValue(const char * key, bool & value, bool required) {
+  VPackSlice tempSlice;
+
+  try {
+    if (_body.isObject() && _body.hasKey(key)) {
+      tempSlice = _body.get(key);
+      value = tempSlice.getBool();
+    } else if (required) {
+      if (_valid) {
+        _result.add(VPackValue(VPackValueType::Object));
+        _valid = false;
+      }
+      _result.add(key, VPackValue("parameter required"));
+    } // else if
+  } catch(VPackException const &vexcept) {
+    if (_valid) {
+      _result.add(VPackValue(VPackValueType::Object));
+      _valid = false;
+    }
+    _result.add(key, VPackValue(vexcept.what()));
+  };
+
+} // RocksDBHotBackup::getParamValue (bool)
+
+
+void RocksDBHotBackup::getParamValue(const char * key, unsigned & value, bool required) {
+  VPackSlice tempSlice;
+
+  try {
+    if (_body.isObject() && _body.hasKey(key)) {
+      tempSlice = _body.get(key);
+      value = tempSlice.getUInt();
+    } else if (required) {
+      if (_valid) {
+        _result.add(VPackValue(VPackValueType::Object));
+        _valid = false;
+      }
+      _result.add(key, VPackValue("parameter required"));
+    } // else if
+  } catch(VPackException const &vexcept) {
+    if (_valid) {
+      _result.add(VPackValue(VPackValueType::Object));
+      _valid = false;
+    }
+    _result.add(key, VPackValue(vexcept.what()));
+  };
```

Comment:
This will work in 99.9999% of all cases, maybe forever, until there is an out of memory situation in one of the `_result.add` calls. Then an exception is thrown and the `_result` object is confused. On the other hand, we cannot really go on anyway in this case, so it is probably not worthwhile to deal with this case.